tensorboard/uploader/proto/write_service.proto

```diff
@@ -119,6 +124,125 @@ message WriteScalarResponse {
   // This is empty on purpose.
 }
 
+// The blob sequence writing procedure works as follows:
+//  * The caller requests creation of a new BlobSequence, via the
+//    CreateBlobSequence RPC. The response provides a new BlobSequence ID.
+//  * The caller writes each blob in the sequence separately via the WriteBlob
+//    RPC as described below, providing the blob sequence ID and sequence index
+//    for each blob.
+//  * Multiple WriteBlob streams may operate in parallel, with no constraint
+//    on their ordering.
+
+// Blob sequence uploads are resumable, as follows:
+//  * The client starts the upload process with a new CreateBlobSequence RPC.
+//  * The blob sequences are uniquely keyed by experiment, run, tag, and step.
+//    Since this is a second (or later) attempt regarding an existing sequence,
+//    a new sequence is not created.  Instead the existing sequence ID is
+//    returned.
+//  * The client attempts to upload each blob in the sequence.  These uploads
+//    resume or may be skipped entirely, as needed, according to the below
+//    procedure.
+
+// The blob writing procedure is as follows.  It is intrinsically resumable.
+//  * The caller instantiates a WriteBlob RPC, which is a bidirectional stream.
+//  * The first of these requests must provide write_blob_first_request_data,
+//    and the subsequent ones must not.
+//  * The first request should send no actual data; it acts as a request for the
+//    current status of the blob, in case it has already been partially or fully
+//    uploaded.
+//  * The first WriteBlobResponse reports the state, size, and crc32c of the
+//    object so far, i.e. taking into account any previously uploaded data.
+//  * The client proceeds to upload from the specified offset (optionally
+//    validating the crc32c up to that point first).
+//  * The caller sends a number of additional WriteBlobRequests, providing the
+//    blob data in an ordered sequence of chunks.
+//  * The caller receives a sequence of WriteBlobResponses in return.  Each
+//    WriteBlobResponse corresponds to one WriteBlobRequest in order, reporting
+//    the state of the write operation.
+//  * In general, the caller should not wait to verify that each response
+//    reports the expected status, size, and crc32c before sending the next
```

Comment:
FWIW, verifying the status and size is cheap, it's only the crc32c verification that is expensive.  If we don't expect the clients to use that field, is it worth setting it on the server?  I forget if we get it for free from GCS.

Reply:
If the field isn't always set on responses, for correctness we should have an explicit indicator to mark it as unset rather than leaving it at the default value of 0, since 0 is a valid CRC output: https://stackoverflow.com/a/49804094/1179226

The most idiomatic way to do that would be using the UInt32Value wrapper type (the well-known wrappers have special handling in JSON interop) and then we can leave it unset if no CRC was provided by the server:
https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/wrappers.proto

That said, up to you but I'd personally be inclined to leave it out unless we're confident that the server will populate it and that we think it's valuable for the client to check it (even in a deferred fashion).  As mentioned in your comment lower in the file, we have robust protection already from the full-file and per-chunk CRCs.

And actually, thinking about it just a bit further, I'm not sure that maintaining a running CRC provides any additional value over the per-chunk CRCs, because if a per-chunk CRC fails to detect corruption in that chunk, I think necessarily any further CRC calculations are not going to make a difference.  See the discussion in https://stackoverflow.com/a/23050676/1179226 which explains that for strings `a` and `b` you can calculate `crc(a + b)` using only `crc(a)`, `crc(b)` and `len(b)`.  So IIUC then if there's a chunk `b'` such that `crc(b') = crc(b)` and `len(b') = len(b)`, then we already know that `crc(a + b') = crc_combine(crc(a), crc(b'), len(b')) = crc_combine(crc(a), crc(b), len(b)) = crc(a + b)`, so the running CRC would also be identical.

I'm definitely not a CRC expert so maybe this analysis is flawed or there are ways to avoid that issue, but given that it seems like even if we did calculate it's possibly doing a lot of extra work for no extra protection, my inclination would be to omit it for now and we can always add it later.